File: implementations/httpcat-deno/cmd.ts
import yargs from "https://deno.land/x/yargs@v17.7.2-deno/deno.ts";

import { fetchPetCodeImage, PetType } from "./logic.ts";
import { convertImageToAscii } from "./logic.ts";
import { isStatusCodeValid } from "./logic.ts";

export async function runCmd() {
  const args = yargs(Deno.args)
    .usage("Usage: $0 status_code [--dog] [--width image_width]")
    .positional("status_code", {
      describe: "HTTP status code",
    })
    .option("dog", {
      type: "boolean",
      describe: "Use if you prefer dogs",
    })
    .option("width", {
      type: "number",
      describe: "Image width",
      default: 100,
    })
    .demandCommand(1)
    .parse();

  let petType = PetType.Cat;
  if (args.dog) {
    petType = PetType.Dog;
  }

  const statusCode = args._[0];
  if (!isStatusCodeValid(statusCode)) {
    console.error(`Invalid status code: ${statusCode}`);
    Deno.exit(1);
  }

  const imageData = await fetchPetCodeImage(petType, args._[0]);
  console.log(await convertImageToAscii(imageData, args.width));
}

File: implementations/httpcat-dart/lib/cmd.dart
import 'dart:io';

import 'package:args/args.dart';
import 'package:httpcat_dart/logic.dart';

void showHelp() {
  print("usage: httpcat-dart HTTP_STATUS_CODE [--dog] [--width WIDTH]");
  print("");
  print("Options:");
  print("  --dog          Use if you prefer dogs");
  print("  --width WIDTH  Image width");
}

void runCmd(List<String> arguments) async {
  final parser = ArgParser();
  parser.addFlag("help", abbr: 'h', help: "Show usage help", defaultsTo: false);
  parser.addFlag("dog",
      help: "Use if you prefer dogs", negatable: false, defaultsTo: false);
  parser.addOption("width", help: "Image width", defaultsTo: "125");

  final results = parser.parse(arguments);
  if (results.rest.isEmpty || results.rest.length > 1 || results.flag("help")) {
    showHelp();
    exit(1);
  }

  final statusCode = int.tryParse(results.rest.first);
  if (statusCode == null) {
    print("ERROR: status code is not a number: ${results.rest.first}\n");
    showHelp();
    exit(1);
  }

  if (!isStatusCodeValid(statusCode)) {
    print("ERROR: status code is not valid: $statusCode\n");
    showHelp();
    exit(1);
  }

  final width = int.tryParse(results.option("width")!);
  if (width == null) {
    print("ERROR: width is not a number: ${results.option("width")}\n");
    showHelp();
    exit(1);
  }

  var petType = PetType.cat;
  if (results.flag("dog")) {
    petType = PetType.dog;
  }

  final imageData = await fetchPetCodeImage(petType, statusCode);
  print(convertImageToAscii(imageData, width));
}
